### DISCUSSION TITLE
My Pallene build and install experience

### DISCUSSION THREAD

**Discussion by bentxt** [pallene-lang/pallene#627] [General]
Hi
I've already red about Pallene a few time, but today I finally tried to build and install the thing on my iMac. It was a bit of a bumpy road but in the end it did work and I can actually run the examples - yay.

With this particular installation routine I wanted to make sure to have everything separated from my standard lua and luarocks setup.

So I'd like to share my experience because on the one hand, I'd like to hear if I made any mistakes and one the other hand this may save time for another newcomer like me.

### User directory and the prefix path
```
# create the work directory, choose what you want
mkdir -p ~/build/pallene

cd ~/build/pallene

# create a directory that can be used as PREFIX for follow up installations, I chose the folder name 'lua',
mkdir lua

# the prefix for later configs in Makefiles etc
export PREFIX=$HOME/build/pallene/lua
```

#### clone the modified lua interpreter
```
cd ~/build/pallene
https://github.com/pallene-lang/lua-internals lua-internals.git
cd lua-internals.git
```

#### edit Makefile of lua-internals
```
from:
INSTALL_TOP= /usr/local
to:
INSTALL_TOP= $(PREFIX)
```
no whitespace after `$(PREFIX)`

and then install with `make guess -j4 && make install`, sudo is not needed

### Installing pallene-tracer

```
cd ~/build/pallene
git clone https://github.com/pallene-lang/pallene-tracer pallene-tracer.git
cd pallene-tracer.git
```
#### edit Makefile of pallene-tracer
```
delete this line:
PREFIX = /usr/local

and change the other line from:
LUA_PREFIX = /usr
to:
LUA_PREFIX= $(PREFIX)
```
no whitespace after `$(PREFIX)`

and then install with `make && make install`, sudo is not needed

### Installing Luarocks

This comes from the Pallene Readme, but the ./configure line is slightly altered, and again sudo is not needed

```
wget https://luarocks.org/releases/luarocks-3.9.1.tar.gz
tar xf luarocks-3.9.1.tar.gz
cd luarocks-3.9.1
./configure --with-lua=$PREFIX --prefix=$PREFIX
make
make install
```

### Installing Pallene

```
cd ~/build/pallene
git clone https://github.com/pallene-lang/pallene pallene.git
cd pallene.git
```

And then the luarocks command slightly altered: 
```
$PREFIX/bin/luarocks make pallene-dev-1.rockspec PTRACER_DIR=$PREFIX
```

### Final Installation result

All the requirements and also the the Pallene compiler itself is installed in $PREFIX, which is in my case ~/build/pallene/lua

In order for the `pallenec` and the c compiler (gcc) compiler to work properly, lets define the include dir:
```
export CFLAGS='-I$PREFIX/include'
```

Lets run the example

```
cd ~/build/pallene/pallene.git
$PREFIX/bin/pallenec  examples/arithmetic/arithmetic.pln
# this results in a .so file in examples/arithmetic/arithmetic.so
$PREFIX/bin/lua  examples/arithmetic/main.lua

1 + 2 = 3
1.5 - 3.25 = -1.75
```
So thats it, comments are welcome

**Comment by bjornbm** [comment]
Thanks for the detailed write-up!

See also the recent `install.sh` in #625.

I've used https://github.com/oploadk/localua(mentioned recently in #620 to build a separated patched lua and pallene on macOS. See https://blog.separateconcerns.com/2022-10-10-localua-pallene.html. But it doesn't support pallene-tracer, I guess.
